### DISCUSSION TITLE
Switching to a better integration method

### DISCUSSION THREAD

**Discussion by multun** [OpenRailAssociation/osrd#138] [Ideas]
The current integration method has a number of issues:
 1) the current method doesn't allow anticipating future directive changes: we can only react by comparing the current state of the train to what it's supposed to be
 2) as we integrate over time and speed controllers work over space, we can only gather information about the starting point of the integration step: we don't know where it ends
 3) the current method generates a lot of data, as there are a lot of integration steps

**Comment by multun** [comment]
I believe the following changes will improve things considerably:

 1) Switch to a higher order method: speed controllers should not only provide speed directives, but also acceleration directives
 2) integrate over space
 3) get variable step integration

**Comment by axrolld** [comment]
I agree switching to an integration over space rather than time for a few reasons :
- the objects we use for running calculation are Maps <Distance, Time>, or <Distance, Speed>, where the key is always the distance, so integrating over distance will ensure two Maps to be comparable point by point
- we usually plot graphs with Distance being the X axis
- this represents better what is being done in real life : a train driver will always react depending on WHERE he/she is, not WHEN, except for departure time
